<a href="https://colab.research.google.com/github/RaheemKProjects/Raheem-North-Star-Analytical-Workflow/blob/main/05_mongodb_optimisation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<a href="https://colab.research.google.com/github/[USERNAME]/northstar-analytics-cw1/blob/main/notebooks/05_mongodb_optimisation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Notebook 05 — MongoDB Atlas: Indexing and Query Optimisation

**NorthStar Analytics — Section 7**

Build indexes on the service_cases collection and verify performance improvement using `explain('executionStats')`.

In [57]:
!pip install pymongo dnspython -q

from pymongo import MongoClient, ASCENDING, DESCENDING
from datetime import datetime
from google.colab import userdata

# Connect to NorthStarDatabase
MONGO_URI = 'mongodb+srv://Raheem:Raheem123@cluster0.rlzqi50.mongodb.net/'
client = MongoClient(MONGO_URI, tlsAllowInvalidCertificates=True)
db     = client['NorthStarDatabase']
cases  = db['service_cases']

print(f'Connected: MongoDB v{client.server_info()["version"]}')
print(f'Cases in collection: {cases.count_documents({})}')
print('Existing indexes:')
for idx in cases.list_indexes():
    print(f'  {idx["name"]}: {dict(idx["key"])}')

Connected: MongoDB v8.0.23
Cases in collection: 886
Existing indexes:
  _id_: {'_id': 1}


Now that we have data, you can re-run the cells for 'Baseline — query without any custom index' and 'Build the index plan' to observe their effects.

## 1. Baseline — query without any custom index

In [58]:
# Drop all custom indexes first — leave only _id
for idx in cases.list_indexes():
    if idx['name'] != '_id_':
        cases.drop_index(idx['name'])

print('Custom indexes dropped.')
print(f'Indexes remaining: {[i["name"] for i in cases.list_indexes()]}')

# Baseline explain on most critical query
baseline = db.command('explain',
    {'find'  : 'service_cases',
     'filter': {'route.pickup_zone': 'Central',
                'delivery.status'  : 'Failed'}},
    verbosity='executionStats')

plan  = baseline['queryPlanner']['winningPlan']
stats = baseline.get('executionStats', {})

print(f'\n--- BASELINE (no indexes) ---')
print(f'Stage            : {plan["stage"]}')
print(f'totalDocsExamined: {stats.get("totalDocsExamined", "N/A")}')
print(f'totalDocsReturned: {stats.get("totalDocsReturned", "N/A")}')
print(f'executionMillis  : {stats.get("executionTimeMillis", "N/A")}')

docs_returned = cases.count_documents({
    'route.pickup_zone': 'Central',
    'delivery.status'  : 'Failed'
})
print(f'totalDocsReturned: {docs_returned}')

baseline_examined = stats.get('totalDocsExamined', 0)
baseline_ms       = stats.get('executionTimeMillis', 0)

Custom indexes dropped.
Indexes remaining: ['_id_']

--- BASELINE (no indexes) ---
Stage            : COLLSCAN
totalDocsExamined: 886
totalDocsReturned: N/A
executionMillis  : 0
totalDocsReturned: 33


Expected baseline: **COLLSCAN**, totalDocsExamined ≈ 886, totalDocsReturned ≈ 30.

## 2. Build the index plan

In [59]:
# Index 1 — compound: zone + status (most operationally critical)
cases.create_index([('route.pickup_zone', ASCENDING),
                    ('delivery.status', ASCENDING)],
                   name='idx_zone_status', background=True)

# Index 2 — customer lookup
cases.create_index([('customer.customer_id', ASCENDING)],
                   name='idx_customer_id', background=True)

# Index 3 — service segment analytics
cases.create_index([('service_type', ASCENDING)],
                   name='idx_service_type', background=True)

# Index 4 — driver scorecards
cases.create_index([('delivery.driver_ref', ASCENDING)],
                   name='idx_driver_ref', background=True)

# Index 5 — multikey on incidents.severity (array)
cases.create_index([('incidents.severity', ASCENDING)],
                   name='idx_inc_severity', background=True)

# Index 6 — recent-cases dashboards
cases.create_index([('created_at', DESCENDING)],
                   name='idx_created_desc', background=True)

# Index 7 — covered query for rating dashboards
cases.create_index(
    [('route.pickup_zone', ASCENDING),
     ('delivery.status', ASCENDING),
     ('delivery.customer_rating', ASCENDING)],
    name='idx_zone_status_rating_cov', background=True)

print('Built indexes:')
for idx in cases.list_indexes():
    print(f'  {idx["name"]}: {dict(idx["key"])}')

Built indexes:
  _id_: {'_id': 1}
  idx_zone_status: {'route.pickup_zone': 1, 'delivery.status': 1}
  idx_customer_id: {'customer.customer_id': 1}
  idx_service_type: {'service_type': 1}
  idx_driver_ref: {'delivery.driver_ref': 1}
  idx_inc_severity: {'incidents.severity': 1}
  idx_created_desc: {'created_at': -1}
  idx_zone_status_rating_cov: {'route.pickup_zone': 1, 'delivery.status': 1, 'delivery.customer_rating': 1}


## 3. Post-index measurement

In [60]:
# Post-index explain using db.command()
optimised = db.command('explain',
    {'find'  : 'service_cases',
     'filter': {'route.pickup_zone': 'Central',
                'delivery.status'  : 'Failed'}},
    verbosity='executionStats')

plan2  = optimised['queryPlanner']['winningPlan']
stats2 = optimised.get('executionStats', {})

def find_index_name(node):
    if 'indexName' in node: return node['indexName']
    if 'inputStage' in node: return find_index_name(node['inputStage'])
    return 'none'

docs_examined = stats2.get('totalDocsExamined', 0)
exec_ms       = stats2.get('executionTimeMillis', 0)

print(f'--- POST INDEX ---')
print(f'Stage            : {plan2["stage"]}')
print(f'Index used       : {find_index_name(plan2)}')
print(f'totalDocsExamined: {docs_examined}')
print(f'totalKeysExamined: {stats2.get("totalKeysExamined", "N/A")}')
print(f'totalDocsReturned: {stats2.get("totalDocsReturned", "N/A")}')
print(f'executionMillis  : {exec_ms}')
print()
print(f'--- IMPROVEMENT ---')
if baseline_examined > 0:
    reduction = round((1 - docs_examined/baseline_examined)*100, 1)
    print(f'Docs examined : {baseline_examined} -> {docs_examined} ({reduction}% reduction)')
print(f'Exec time (ms): {baseline_ms} -> {exec_ms}')

--- POST INDEX ---
Stage            : FETCH
Index used       : idx_zone_status
totalDocsExamined: 33
totalKeysExamined: 33
totalDocsReturned: N/A
executionMillis  : 2

--- IMPROVEMENT ---
Docs examined : 886 -> 33 (96.3% reduction)
Exec time (ms): 0 -> 2


## 4. Covered query — fastest possible read pattern

When the index includes every field used in the filter and the projection, MongoDB returns the result entirely from the index without fetching any documents.

In [61]:
# Covered query using db.command()
covered = db.command('explain',
    {'find'      : 'service_cases',
     'filter'    : {'route.pickup_zone': 'Central',
                    'delivery.status'  : 'Failed'},
     'projection': {'delivery.customer_rating': 1, '_id': 0}},
    verbosity='executionStats')

plan3 = covered['queryPlanner']['winningPlan']

def walk(node, indent=0):
    print('  ' * indent + node['stage'] +
          (f' (index: {node.get("indexName")})' if node.get('indexName') else ''))
    if 'inputStage' in node:
        walk(node['inputStage'], indent + 1)

print('Stage chain:')
walk(plan3)
docs3 = covered.get('executionStats', {}).get('totalDocsExamined', 'N/A')
print(f'\ntotalDocsExamined: {docs3}')
print('(0 = covered query confirmed — answered from index only)')

Stage chain:
PROJECTION_DEFAULT
  IXSCAN (index: idx_zone_status_rating_cov)

totalDocsExamined: 0
(0 = covered query confirmed — answered from index only)


## 5. Aggregation pipeline optimisation

Three patterns: $match early, $project to drop unused fields, $sort backed by an index.

In [62]:
from datetime import datetime

optimised_pipeline = [
    # Rule 1: $match FIRST to reduce working set
    {'$match': {
        'delivery.status': {'$in': ['Failed', 'Delayed']}
    }},
    # Rule 2: $project to drop unused fields
    {'$project': {
        'route.pickup_zone'       : 1,
        'delivery.status'         : 1,
        'delivery.customer_rating': 1
    }},
    # Rule 3: $group
    {'$group': {
        '_id'       : '$route.pickup_zone',
        'n_problem' : {'$sum': 1},
        'avg_rating': {'$avg': '$delivery.customer_rating'}
    }},
    {'$addFields': {
        'avg_rating': {'$round': ['$avg_rating', 2]}
    }},
    # Rule 4: $sort last
    {'$sort': {'n_problem': -1}}
]

print('Optimised pipeline results:')
for d in cases.aggregate(optimised_pipeline):
    print(f"  Zone: {d['_id']:12} | problems: {d['n_problem']:4} | avg_rating: {d['avg_rating']}")

Optimised pipeline results:
  Zone: Central      | problems:   84 | avg_rating: 2.88
  Zone: East         | problems:   50 | avg_rating: 3.12
  Zone: North        | problems:   43 | avg_rating: 3.09
  Zone: Airport      | problems:   43 | avg_rating: 3.41
  Zone: Riverside    | problems:   43 | avg_rating: 3.17
  Zone: South        | problems:   36 | avg_rating: 3.16
  Zone: West         | problems:   35 | avg_rating: 3.01


## 6. Index utilisation report

Per-index access statistics from MongoDB's $indexStats — useful for retiring indexes that are not earning their keep.

In [63]:
for stat in cases.aggregate([{'$indexStats': {}}]):
    print(f"{stat['name']:35s}  ops: {stat['accesses']['ops']:6d}"
          f"  since: {stat['accesses']['since']}")

idx_created_desc                     ops:      0  since: 2026-05-12 17:03:50.675000
idx_inc_severity                     ops:      0  since: 2026-05-12 17:03:50.581000
idx_zone_status_rating_cov           ops:      0  since: 2026-05-12 17:03:50.745000
idx_service_type                     ops:      0  since: 2026-05-12 17:03:50.423000
idx_driver_ref                       ops:      0  since: 2026-05-12 17:03:50.506000
_id_                                 ops:      0  since: 2026-05-12 16:53:07.217000
idx_zone_status                      ops:      0  since: 2026-05-12 17:03:50.283000
idx_customer_id                      ops:      0  since: 2026-05-12 17:03:50.352000
